erts/emulator/test/nif_SUITE.erl

```diff
@@ -2610,6 +2614,81 @@ nif_snprintf(Config) ->
     <<"{{hello,world,-33},",0>> = format_term_nif(20,{{hello,world, -33}, 3.14, self()}),
     ok.
 
+nif_phash2(Config) ->
+    ensure_lib_loaded(Config),
+    Terms =
+        [random_term() || _ <- lists:seq(1, 1000)],
+
+    lists:foreach(
+      fun (Term) ->
+              HashValue = erlang:phash2(Term),
+              NifHashValue = phash2_nif(Term),
+              (HashValue =:= NifHashValue
+               orelse ct:fail("Expected: ~p\nActual:   ~p",
+                              [HashValue, NifHashValue]))
+      end,
+      Terms).
+
+nif_phash2_ranged(Config) ->
+    ensure_lib_loaded(Config),
+    RandomRangedTerms =
+        [{random_term(), rand:uniform((1 bsl 32) - 1)}
+         || _ <- lists:seq(1, 1000)],
+
+    lists:foreach(
+      fun ({Term, Range}) ->
+              HashValue = erlang:phash2(Term, Range),
+              NifHashValue = phash2_ranged_nif(Term, Range),
+              (HashValue =:= NifHashValue
+               orelse ct:fail("Expected: ~p\nActual:   ~p",
+                              [HashValue, NifHashValue]))
+      end,
+      RandomRangedTerms),
+
+    EdgeCaseTerm = random_term(),
+    EdgeCaseHashValue = erlang:phash2(EdgeCaseTerm, 1 bsl 32),
+    EdgeCaseNifHashValue = phash2_ranged_nif(EdgeCaseTerm, 0),
+    (EdgeCaseHashValue =:= EdgeCaseNifHashValue
+     orelse ct:fail("Expected: ~p\nActual:   ~p",
+                    [EdgeCaseHashValue, EdgeCaseNifHashValue])).
+
+-define(HALF_DBL_EPSILON, 1.1102230246251565e-16). % math:pow(2, -53)
+
+random_term() ->
+    case rand:uniform(6) of
+        1 -> rand:uniform(1 bsl 27) - 1; % small
+        2 -> (1 bsl 27) + rand:uniform(1 bsl 128); % big
+        3 -> random_sign() * (rand:uniform() * ?HALF_DBL_EPSILON); % float
```

Comment:
This way of generating random floats doesn't make sense, I'll fix it when I get a chance.